Implement issue #68 - add pointer_mut

CryptArchy opened the pull request:
I was thinking that JSON Patch support would be useful, and saw the Issues for that and for `pointer_mut` support. #68 `pointer_mut` seemed necessary for implementing Patch, so I took a swing at it! This is my first open-source Rust PR so criticism is welcome but be gentle, haha!

I moved `pointer` into a new trait (`Pointer`) and added `pointer_mut` to it as well.  To maintain symmetry I also added a `pointer_owned` method that consumes the `Value` and returns an owned `Value`. Because of the new trait this is a breaking change (you need to import the trait for `pointer` to work now.)

Unit tests were added and all pass, documentation was added but is mostly just copy-paste from the existing code. I also wrote a benchmark to verify that the changes I made to `pointer` weren't slower than the existing code, but left out the change to the `.rs.in` file I made to run the bench. On my machine, the new and old `pointer` methods were very close in performance. The new one actually seems slightly faster but they were within deviation range of each other so I'm making no claims except that my refactors shouldn't have negatively affected the method.

oli-obk commented:
Why did you move everything into a trait? Do you have a use case for implementing those methods on another type?

Don't worry about the failing travis, it's just nightly and not related to serde directly.

Do you think it would be feasible to implement those methods through a macro so there's only one impl written?

oli-obk commented on json/src/pointer.rs:
```diff
@@ -0,0 +1,141 @@
+            let target_opt = match target {
+                Value::Object(mut map) => map.remove(&token[..]),
+                Value::Array(mut list) => {
+                    parse_index(&token[..]).and_then(|x| Some(list.remove(x)))
```
This should probably be `map` instead of and_then

CryptArchy commented:
Corrected the `map` instead of `and_then`.  Good catch, thank you!

I was using a trait to keep things clean while I worked, but now that I think about it more it does seem odd. If made generic over the return value (instead of being `serde_json::Value`) then the trait could generically represent any means of accessing elements via strings, like for XPath maybe.  But I acknowledge that's stretching things a bit.

It's no problem for me to move it out of the trait and put the methods into the Value impl as before.  

I'm not sure I understand what you mean about a macro.  Like doing a `pointer!(v: Value, s: &str)` kind of thing and have the macro be generic in the return based on the ref-type of the parent Value?

oli-obk commented on json/src/pointer.rs:
```diff
@@ -0,0 +1,141 @@
+use value::Value;
+
+/// Provides the `pointer` method for locating values within an object using a string path.
+pub trait Pointer {
```
Yes, please move these methods directly onto the type. If we ever get a more concrete use case for splitting them off, we'll do that then

CryptArchy replied:
Done.

oli-obk commented on json/src/pointer.rs:
```diff
@@ -0,0 +1,141 @@
+                Value::Object(ref mut map) => map.get_mut(&token[..]),
+                Value::Array(ref mut list) => {
+                    if let Some(idx) = parse_index(&token[..]) {
+                        list.get_mut(idx)
```
For symmetry with the other methods, also use `map` here

CryptArchy replied:
Done.  I had originally had trouble with the closure, but it wasn't hard to correct (just needs a `move`) once I worked out the rest of the borrow checker's nagging. 😃

oli-obk commented on json/src/pointer.rs:
```diff
@@ -0,0 +1,141 @@
+            let target_opt = match *target {
+                Value::Object(ref map) => map.get(&token[..]),
+                Value::Array(ref list) => {
+                    parse_index(&token[..]).and_then(|x| list.get(x))
```
I know this was in there before, but can you check whether it works to just pass `token` instead of `&token[..]`

CryptArchy replied:
It won't take `token` but it will take just `&token`.  I wasn't sure why the full range slice was in there, I just assumed there was some subtlety I was missing.

CryptArchy commented:
Ah ok, I originally thought you had meant something like that, and I had considered it.  At the moment, my brain isn't making the leap to how I would write it (or maybe I'm just not very good at macros yet!) The code does seem so very similar that it feels like some abstraction should work there, but the changes for each function are subtle and pervasive so it'd be tricky.

oli-obk commented on json/src/value.rs:
```diff
@@ -161,26 +168,97 @@ impl Value {
+    /// For more information read [RFC6901](https://tools.ietf.org/html/rfc6901).
+    pub fn pointer_owned(self, pointer: &str) -> Option<Value> {
+        if pointer == "" {
+            // except for this, the function could probably be `&mut self` and only consume the child element
```
that is a good point!

@dtolnay what do you think about using `&mut self` and turning it into `Value::Null` in the case of `pointer == ""`?

dtolnay replied:
I would prefer to keep it the way @CryptArchy has written it. The behavior you suggested is `v.pointer_mut(p).map(|m| mem::replace(m, Value::Null))` which makes it clear in the caller's code exactly how the value is being mutated.

dtolnay commented:
Please remove the carriage return characters from test_json.rs. That is why the file looks like it has 3396 lines of diff.

CryptArchy commented:
Oh, just noticed that. I had to fix a merge conflict and the diff tool apparently loves to mess up line endings. Sorry about that.

dtolnay commented:
:+1: Looks good to me. I'll give @oli-obk the final say because he has been more involved in this PR.

CryptArchy commented:
So I tested out what @dtolnay was saying and it looks like his little one liner is enough to completely replace the `point_owned` behavior.  As in:

```rust
pub fn pointer_owned(&mut self, pointer: &str) -> Option<Value> {
        use std::mem;
        self.pointer_mut(pointer).map(|m| mem::replace(m, Value::Null))
}
```

So that's pretty cool! I forget about the `mem` library, should probably use it more often. Should I push that version up, or since it's so simple do we want to even keep the `pointer_owned` version anymore?

CryptArchy commented:
As an aside, the `pointer_owned` method I said was for symmetry, but it was actually an experiment.  I implemented it first to better understand the ownership behavior around the Value graph, as it was a lot easier than the mutable case. Then I just thought it seemed useful to leave in, hehe.

CryptArchy commented:
Sorry, lots of babbling on my part. @oli-obk if you want me to change or remove `pointer_owned` then I'm fine either way.  Or if you like the current implementation then it can stay as is, of course. Up to you, just let me know.

oli-obk commented:
Yea, I'd say we remove it without an explicit use case, because I think it's problematic to move out the entire value object just to extract a single thing by value. All non-nulling designs I can think of return both the value and the "rest" (if there is any rest), but none are ergonomic.

CryptArchy commented:
Alright, removed `pointer_owned` and added a little more tests and documentation to cover David's neat little `mem::replace` trick and otherwise show how to use `pointer_mut`.

oli-obk commented:
Wonderful, thanks for being so thorough

CryptArchy commented:
My pleasure, thanks for the code reviews and being patient while I fixed things up. Looking forward to the next one, haha.
